### ISSUE TITLE
[Feature] Use existing vite.config.js file for component tests

### ISSUE THREAD

**Issue by craigjennings11** [microsoft/playwright#14295]
I have a project that already uses Vite, and is configured with the [vite-tsconfig-paths](https://github.com/aleclarson/vite-tsconfig-paths) plugin to resolve my paths correctly. It looks like Playwright's new component testing feature doesn't use my existing `vite.config.js` file as it errors when trying to resolve a path that needs to follow my tsconfig path resolution logic.

It'd be nice if Playwright could make use of my custom config and just extend it to point to the correct `entry` point (e.g. `playwright/index.html`), or at least allow for configuring extra plugins for Playwright's vite config to work correctly.

**Comment by rwoll** [comment]
It's not yet documented[^1], but you can explicitly configure the config (along with a few other options) via: https://github.com/microsoft/playwright/blob/fbb364c1cd0dec9e4a75b885d51e67ca397b1d6a/packages/playwright-ct-react/index.d.ts#L33

```ts
import type { PlaywrightTestConfig } from '@playwright/experimental-ct-react';
import { devices } from '@playwright/experimental-ct-react';

/**
 * See https://playwright.dev/docs/test-configuration.
 */
const config: PlaywrightTestConfig = {
  // …
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',

    /* Port to use for Playwright component endpoint. */
    ctPort: 3100,

    /* Path to custom config or inline config */
    ctViteConfig: './path/to/config/…',
  },

  /* Configure projects for major browsers */
  projects: [
    // …
  ],
};

export default config;
```

Give it a try and let us know how it goes! If you encounter any issues or have questions, feel free to comment here or open up another issue.

[^1]: Since we announced the feature as a preview, we have not yet given the docs the ❤️ they deserve…yet!

**Comment by pavelfeldman** [comment]
For now, we are bundler-agnostic, so we are not reusing your existing config. Your config might have a lot of things we won't be able to use. So for now, you would copy your path mappings using ctViteConfig. In the future, we will try to provide a solution that interprets your vite config and reuses relevant bits from it!

**Comment by pavelfeldman** [comment]
Folding into https://github.com/microsoft/playwright/issues/14298

**Comment by craigjennings11** [comment]
@rwoll That worked great! Thanks for pointing that out.

**Comment by ScubaDaniel** [comment]
> Give it a try and let us know how it goes! If you encounter any issues or have questions, feel free to comment here or open up another issue.

It seems like `ctViteConfig` is being ignored completely? I'm trying to fix an issue with Vite based on monorepo symbols ([like this](https://github.com/vitejs/vite/issues/5668)).

I've been trying all the documented work-arounds, but it seems like Vite/Rollup aren't getting my configuration - at least the ones that matter.

Things that don't seem to work:
* optimizeDeps
* build.commonjsOptions
* build.rollupOptions

On a side note, why was Vite chosen over another bundler like webpack? Vite has been giving our team lots of headache due to various issues that webpack doesn't have.

**Comment by AlexanderMykulych** [comment]
> 

Had the same problem, some of my configurations applied, and some did not. I've written my custom vite plugin and defined there `config` function:
<img width="238" alt="image" src="https://user-images.githubusercontent.com/7901909/172836278-584d6ea3-7562-49e6-8dcb-0dc8534d2ce2.png">

<img width="250" alt="image" src="https://user-images.githubusercontent.com/7901909/172836200-979003f1-5b04-45f6-b484-5c663223be4c.png">

 It helped.

**Comment by WesleyYue** [comment]
Getting this working in a monorepo was so insanely difficult but finally did it.

> It seems like certain ctViteConfig is being ignored?

@ScubaDaniel This is because Playwright is [overwriting entire sections of the config](https://github.com/microsoft/playwright/blob/b695478840390a3974196c7578775d05fbc87a15/packages/playwright-test/src/plugins/vitePlugin.ts#L98)

Following Alexander's suggestion to write our own plugin, this is the full example config I have that works for a monorepo with Yarn workspaces, typescript, and commonjs modules

Assumes you have Yarn workspaces that looks like the following:
```
packages/
├─ package-1/
├─ dependency-of-package-1/
```

If `dependency-of-package-1` can be built as esm modules (`module: 'esnext'` in tsconfig.json), then you don't need to tinker with `optimizeDeps` and `build.commonjsOptions` in Vite. If your `dependency-of-package-1` needs to be commonjs, you need to tell Vite to convert the commonjs modules to esmodules, with the following:

```
// playwright-ct.config.ts
import { resolve } from "node:path";
import type { PlaywrightTestConfig } from "@playwright/experimental-ct-react";
import { devices } from "@playwright/experimental-ct-react";
import react from "@vitejs/plugin-react";
import type { UserConfig } from "vite";

function monorepoWorkaround() {
  return {
    name: "monorepo-workaround",
    config(config: UserConfig): void {
      config.optimizeDeps!.include?.push("@company/dependency-of-package-1"); // This is the name of the package in package.json. I prefix our monorepo packages with our own @company, but you don't have to.

      config.build!.commonjsOptions = {
        include: [/dependency-of-package-1/, /node_modules/], // dependency-of-package-1 should be the folder name of your dependency pacakge.
      };
    },
  };
}

/**
 * See https://playwright.dev/docs/test-configuration.
 */
const config: PlaywrightTestConfig = {
  testDir: "./",
  testMatch: /.*.ct-spec.tsx/,
  /* The base directory, relative to the config file, for snapshot files created with toMatchSnapshot and toHaveScreenshot. */
  snapshotDir: "./__snapshots__",
  /* Maximum time one test can run for. */
  timeout: 10 * 1000,
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: Boolean(process.env.CI),
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: "html",
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: "on-first-retry",

    /* Port to use for Playwright component endpoint. */
    ctPort: 3100,

    ctViteConfig: {
      plugins: [monorepoWorkaround(), react()],
      define: {
        "process.env": process.env,  // Port over REACT_APP envs
      },
      resolve: {
        alias: {
          "@": resolve(__dirname, "./src"),  // My custom alias
        },
      },
    },
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: "chromium",
      use: {
        ...devices["Desktop Chrome"],
      },
    },
    {
      name: "firefox",
      use: {
        ...devices["Desktop Firefox"],
      },
    },
    {
      name: "webkit",
      use: {
        ...devices["Desktop Safari"],
      },
    },
  ],
};

// eslint-disable-next-line import/no-default-export
export default config;
```

**Comment by FACorreiaa** [comment]
This isn't fixed yet as Im having problems testing Jotai atoms with a vite config
